### ISSUE TITLE
[epic] Isograph VSCode extension

### ISSUE THREAD

**Issue by rbalicki2** [isographlabs/isograph#48]
We need to have a VSCode extension that provides
* syntax highlighting
    * including highlighting fields in different colors based on whether they are server fields or client fields, etc.
    * when there are unwraps, perhaps colors fields based on their type
* error reporting
* go to definition of client and server fields
* autocomplete
* find all uses of client fields/server fields
etc.

**Comment by edmondop** [comment]
I would be interested in this but I am not having all the necessary skills... would love to pair / get a piece of this work

**Comment by rbalicki2** [comment]
Absolutely. Happy to. This is a high priority, and will take a while. It's
probably good to get started now.

The outline I have is:
* more or less, follow the setup of the Relay VSCode LSP:
  * how the "server" dispatch is setup:
[`compiler/crates/relay-lsp/src/server/lsp_notification_dispatch.rs`](
https://github.com/facebook/relay/blob/16e28af78b473f848792d132d3f9f26ecfd37dea/compiler/crates/relay-lsp/src/server/lsp_notification_dispatch.rs#L4),
`on_notification_sync`, `dispatch_notification`, `dispatch_request`, etc.
  * https://docs.rs/lsp-types/latest/lsp_types/ (and
https://github.com/rust-lang/rust-analyzer/tree/master/lib/lsp-server/src ?)
  * Something like
https://github.com/facebook/relay/blob/16e28af78b473f848792d132d3f9f26ecfd37dea/compiler/crates/resolution-path/src/lib.rs,
which in Relay, essentially "inverts" GraphQL structures, and is used to
generate a data structure describing exactly what is underneath a cursor in
a typesafe way: e.g. a scalar field, whose parent is a linked field, whose
parent is an isograph field definition.
* This is my rough ordering of LSP features, from highest to lowest
priority: (top) syntax highlighting, hover information, error display, go
to definition (for client fields), auto complete, go to definition for
server fields, find uses of fields, quick fixes and refactors, status
display in the bottom VSCode bar,

What else do we need to do to actually ship this?
* To actually consider the VSCode extension suitable for public
consumption, I think we would require big refactor of the compiler, which
is currently a purely batch mode compiler. For that, I want to adopt a
Salsa-like architecture. Here's a good [youtube](
https://youtu.be/vrnNvAAoQFk?si=Nfx0dvp55nMfjXaj).
    * However, we can do most of the LSP feature work without waiting on
this.

**Comment by edmondop** [comment]
As discussed in person priorities ares:
- Syntax highlighting (Request)
- Error reporting  (LSPStatusReporter in the Relay LSP)
- Autocomplete/Tab completion
- Hover

**Comment by rbalicki2** [comment]
Closing as there are probably follow up issues
